### PR TITLE
Ensure volume points get created, and with perms

### PR DESCRIPTION
**Pull request by nalind**
Ensure that volume mount points are created, if they don't exist, when they're defined in a Dockerfile (#151), and that if we create them, we create them with 0755 permissions (#152).

### CODE REVIEW

File: imagebuildah/build.go [hunk 1]
```diff
@@ -241,6 +255,9 @@ func (b *Executor) volumeCacheSave() error {
 		if !os.IsNotExist(err) {
 			return errors.Wrapf(err, "error checking for cache of %q in %q", archivedPath, cacheFile)
 		}
+		if err := os.MkdirAll(archivedPath, 0755); err != nil {
```
Comment (rhatdan) [inline imagebuildah/build.go]: Will this work on a READONLY Image?

**Comment by rhatdan** [comment]
Does buildah ignore when a dockerfile attempts to write to the directory or if content is created in the directory.  I guess I am not sure what happens in docker when this happens.

**Comment by rhatdan** [comment]
@runcom @mrunalp PTAL

**Comment by runcom** [comment]
In CRI-O we create those VOLUME(s) with 0644 actually https://github.com/kubernetes-incubator/cri-o/blob/master/server/container_create.go#L610 :confused:

**Comment by rhatdan** [comment]
They should be consistent, although creating a directory as 644 means that you could not search through it, 755, is probably correct.

**Comment by mrunalp** [comment]
I just tested volume behavior with docker:
```
[root@localhost test_volume]# docker build -t test_volume .
Sending build context to Docker daemon 2.048 kB
Step 1 : FROM busybox:latest
 ---> e02e811dd08f
Step 2 : VOLUME /mydata
 ---> Running in c6a07ccebdfd
 ---> 7af942b3d1e1
Removing intermediate container c6a07ccebdfd
Step 3 : RUN echo hi > /mydata/hi.txt
 ---> Running in 0cda24be6cfb
 ---> a893656b0a28
Removing intermediate container 0cda24be6cfb
Successfully built a893656b0a28
[root@localhost test_volume]# docker run -it --rm test_volume sh
/ # ls -l /mydata/
total 0
/ # 
```
So, it looks like docker isn't saving contents written to a volume during a build. This would mean that we would want to use bind mounts here. Otherwise, we cannot mimic this behavior if there already existed a directory in the image that a volume is added over.

**Comment by rhatdan** [comment]
I think in CRI-O they are mounting a tmpfs there now, which would probably work for buildah.  CRI-O is looking to handle it differently since OpenShift wants us to stop users from using storage outside of volumes handed to the user.

**Comment by mrunalp** [comment]
We are also doing a mkdir in CRI-O right now, but I plan to make it all configurable as part of my current work. There could be these options in the config for admin:
1. ImageVolume=ignore
2. ImageVolume=tmpfs
3. ImageVolume=mount
4. ImageVolume=directory

**Comment by nalind** [comment]
@mrunalp Apparently RUN instructions which modify the contents of a directory that's been marked as a volume are discarded, but contents can still be added using ADD or COPY instructions (noted at https://github.com/openshift/imagebuilder/blob/master/builder.go#L213, and experimentally confirmed).

File: imagebuildah/build.go (hunk 1 above)
Comment (nalind) [inline imagebuildah/build.go]: At this point we should be operating on a read-write layer, so I don't expect whether or not the original image is on a writable filesystem to matter.

**Comment by mrunalp** [comment]
@nalind okay, I guess it is hard to support RUN as commands could be arbitrary but with COPY / ADD it is easy to parse the destination and check which would explain the behavior.

**Comment by nalind** [comment]
Updated to use the contents of a volume mount point to initialize a directory that we bind mount in from the host if nothing is specified for a volume location.  This appears to match the behavior of "docker run" and "docker start" for images that contain data in volume directories.

File: run.go [hunk 2]
```diff
@@ -98,17 +100,37 @@ func setupMounts(spec *specs.Spec, optionMounts []specs.Mount, bindFiles, volume
...
+			srcPath := filepath.Join(mountPoint, volume)
+			if err = archive.CopyWithTar(srcPath, volumePath); err != nil {
+				return errors.Wrapf(err, "error populating directory %q for volume %q in container %q using contents of %q", volumePath, volume, b.ContainerID, srcPath)
+			}
+
+		}
+		// Add the bind mount.
 		mounts = append(mounts, specs.Mount{
-			Source:      "tmpfs",
+			Source:      volumePath,
```
Comment (nalind) [inline run.go]: Yes, it is cleaned up by the storage library.

**Comment by rhatdan** [comment]
@rh-atomic-bot r+ ed4cd48

**Pull request closed without merging**